File: sort-array/solution1.py
#!/bin/python3

#
# O(n) solution
#

def sort_binary(data):
  res = [1] * len(data)
  i = 0
  for num in data:
    if num == 0:
      res[i] = num
      i += 1
  return res

# 
# TESTS
# 

data = [1, 0, 1, 0, 1, 0, 0, 1]

res = [0, 0, 0, 0, 1, 1, 1, 1]

res_ = sort_binary(data)

if res_ != res:
  raise(Exception("Expected %s, got %s" % (res, res_)))

File: sort-array/solution1.clj
(defn sort-binary
  [data]
  (let [data* (group-by identity data)]
    (concat (get data* 0) (get data* 1))))

;;
;; TESTS
;;

(let [data [1, 0, 1, 0, 1, 0, 0, 1]
      res [0, 0, 0, 0, 1, 1, 1, 1]
      res* (sort-binary data)]
  (when (not= res res*)
    (throw (Exception. (format "Expected %s got %s" res (doall res*))))))